internal/collector/reconcile/gateway.go

```diff
@@ -0,0 +1,139 @@
+// Copyright The OpenTelemetry Authors
+// Copyright Splunk Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reconcile
+
+import (
+	"context"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/signalfx/splunk-otel-collector-operator/internal/collector"
+	"github.com/signalfx/splunk-otel-collector-operator/internal/naming"
+)
+
+// +kubebuilder:rbac:groups="apps",resources=deployments,verbs=get;list;watch;create;update;patch;delete
+
+// Gateway reconciles the Splunk Otel Gateway required for the instance in the current context.
+func Gateways(ctx context.Context, params Params) error {
+	desired := []appsv1.Deployment{}
+	if params.Instance.Spec.Gateway.Enabled != nil && *params.Instance.Spec.Gateway.Enabled {
+		// TODO(splunk): pass params.Instance.Spec.Gateway instead of params.Instance
+		desired = append(desired, collector.Gateway(params.Log, params.Instance))
+	}
+
+	// first, handle the create/update parts
+	if err := expectedGateways(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the expected deployments: %w", err)
+	}
+
+	// then, delete the extra objects
+	if err := deleteGateways(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the deployments to be deleted: %w", err)
+	}
+
+	return nil
+}
+
+func expectedGateways(ctx context.Context, params Params, expected []appsv1.Deployment) error {
```

Comment:
this is complex enough it warrants unit tests

Reply:
Got point, added.